File: Level-2/reverse_sentence.cc
/*
 * Date: 2018-11-10
 *
 * Description:
 * Given a sentence, reverse it's individual words. For example if sentence is
 * "Hello World" it should be changed to "World Hello".
 *
 * Approach:
 * - First, reverse whole string
 * - Take individual words from reversed string and reverse them
 * - Above 2 steps will result in reversed words in sentence.
 *
 * Complexity:
 * O(N)
 */

#include "iostream"

using namespace std;

string reverse_string(string s) {
  int start = 0;
  int end = s.size() - 1;
  char c;

  while (start < end) {
    c = s[start];
    s[start] = s[end];
    s[end] = c;
    start += 1;
    end -= 1;
  }
  return s;
}

int main() {
  string input;
  string reversed;
  string temp = "";
  string result = "";
  int i = 0;

  cout << "Enter a sentence: ";
  getline(cin, input);

  // Step 1: Reverse whole string
  reversed = reverse_string(input);
  cout << "Reversed string: " << reversed << endl;

  // Step 2: Now reverse individual words
  while (i <= reversed.size()) {
    if (reversed[i] == ' ' || reversed[i] == '\0') {
      result += reverse_string(temp) + ' ';
      temp = "";
    }
    else
      temp += reversed[i];
    i += 1;
  }
  result += temp;
  cout << "Reversed words in sentence: " << result << endl;
}


/*
 * Output:
 * -------------------
 * Enter a sentence: Coding is fun
 * Reversed string: nuf si gnidoC
 * Reversed words in sentence: fun is Coding 
 *
 * Enter a sentence: hello
 * Reversed string: olleh
 * Reversed words in sentence: hello 
 */

File: Level-2/reverse_sentence.py
# Date: 2020-10-01
#
# Description:
# Given a sentence, reverse it's individual words. For example if sentence is
# "Hello World" it should be changed to "World Hello".
#
# Approach:
# - First, reverse whole string
# - Take individual words from reversed string and reverse them
# - Above 2 steps will result in reversed words in sentence.
#
# Complexity:
# O(N)

def _reverse(l):
    start = 0
    end = len(l) - 1
    while start < end:
        l[start], l[end] = l[end], l[start]
        start += 1
        end -= 1
    return l

def reverse_sentence(sentence):
    if not sentence:
        return sentence
    sentence = list(sentence)
    rev = _reverse(sentence)

    i = 0
    temp = []
    res = []
    while i < len(rev):
        if rev[i] == ' ':
            res.extend(_reverse(temp) + [' '])
            temp = []
        else:
            temp.append(rev[i])
        i += 1
    if temp:
        res.extend(_reverse(temp))
    return ''.join(res)


assert reverse_sentence('hello world') == 'world hello'
assert reverse_sentence('world') == 'world'
